invokeai/backend/stable_diffusion/diffusion/custom_attention.py

```diff
@@ -0,0 +1,342 @@
+from invokeai.backend.util.devices import TorchDevice
+
+
+@dataclass
+class IPAdapterAttentionWeights:
+    ip_adapter_weights: IPAttentionProcessorWeights
+    skip: bool
+
+
+class CustomAttnProcessor:
+    """A custom implementation of attention processor that supports additional Invoke features.
+    This implementation is based on
+    SlicedAttnProcessor (https://github.com/huggingface/diffusers/blob/fcfa270fbd1dc294e2f3a505bae6bcb791d721c3/src/diffusers/models/attention_processor.py#L1616)
+    AttnProcessor2_0 (https://github.com/huggingface/diffusers/blob/fcfa270fbd1dc294e2f3a505bae6bcb791d721c3/src/diffusers/models/attention_processor.py#L1204)
+    Supported custom features:
+    - IP-Adapter
+    - Regional prompt attention
+    """
+
+    def __init__(
+        self,
+        ip_adapter_attention_weights: Optional[List[IPAdapterAttentionWeights]] = None,
+    ):
+        """Initialize a CustomAttnProcessor.
+        Note: Arguments that are the same for all attention layers are passed to __call__(). Arguments that are
+        layer-specific are passed to __init__().
+        Args:
+            ip_adapter_weights: The IP-Adapter attention weights. ip_adapter_weights[i] contains the attention weights
+                for the i'th IP-Adapter.
+        """
+
+        self._ip_adapter_attention_weights = ip_adapter_attention_weights
+        self.attention_type, self.slice_size = self._select_attention()
+
+        # inspect didn't work because it's native function
+        # In 2.0 torch there no scale argument in sdp, it's added in 2.1
+        # Probably can selected based on torch version instead
+        try:
+            F.scaled_dot_product_attention(torch.zeros(1,1), torch.zeros(1,1), torch.zeros(1,1), scale=0.5)
+            self.scaled_sdp = True
+        except:
+            self.scaled_sdp = False
+
+    def _select_attention(self):
+        config = get_config()
+        attention_type = config.attention_type
+        if attention_type in ["normal", "xformers"]:
+            logger.warning(f'Attention "{attention_type}" no longer supported, "torch-sdp" will be used instead.')
+            attention_type = "torch-sdp"
+
+        if attention_type == "auto":
+            exec_device = TorchDevice.choose_torch_device()
+            if exec_device.type == "mps":
+                attention_type = "sliced"
+            else:
+                attention_type = "torch-sdp"
+
+        if attention_type == "torch-sdp" and not hasattr(F, "scaled_dot_product_attention"):
+            raise ImportError("torch-sdp attention requires PyTorch 2.0, to use it, please upgrade PyTorch to 2.0.")
+
+        slice_size = None
+        if attention_type == "sliced":
+            slice_size = config.attention_slice_size
+            if slice_size not in ["auto", "balanced", "max"] and not isinstance(slice_size, int):
+                raise ValueError(f"Unsupported attention_slice_size: {slice_size}")
+            if slice_size == "auto":
```

Comment:
More uncertainty about whether the canonical value for `slice_size` is "balanced" or "auto".